Jdbc based database connection

Czeffik opened the pull request.

Czeffik committed:
Add docker-compose with postgres and db setup via flyway

Czeffik committed:
Add jdbc based database interactions

WHATS DONE:
- added AccountDao for CRUD operations

Czeffik committed:
Add transaction example

tomaszc commented:
Hejże, jutro dostaniesz komenty:D

tomaszc commented on readme.md:
```diff
@@ -0,0 +1,9 @@
+* [Docker](https://www.docker.com/get-started)
+* JAVA
+
+## How to start
```
ok, wszystko fajnie ale .... co to robi? jak powinienem wystartować aplikację? co się stanie po jej odpaleniu? jaki jest jej cel?
opisz to wszystko tutaj co chcesz zaprezentować, zrób jakąś funkcjonalność goddamnit! :D

tomaszc commented on src/main/java/com/trzewik/jdbc/App.java:
```diff
@@ -1,4 +1,13 @@
+
 public class App {
+    public static void main(String[] args) throws SQLException {
+        Dao<Account> dao = DbFactory.accountDao();
```
eee i co ? Panie, ja chcę odpalić apikacje i mieć jakieś GUI, daj mi poprzez konsolę i komentarze możliwość wylistowania accountów lub ich dodania,

Czeffik replied:
zrobione....

tomaszc commented on src/main/java/com/trzewik/jdbc/db/Account.java:
```diff
@@ -0,0 +1,25 @@
+import java.sql.SQLException;
+
+@Getter
+@EqualsAndHashCode
```
jestem ciekawy czy jesteś pewny że dobrze określono ten bardzo ważny combo (equals+hash-code) - wyjaśnij czemu taki plz

Czeffik replied:
wcześniej było `@EqualsAndHashCode` na wszystkich polach, ponieważ wszystkie pola były unikalne... ale nie wiem czy to miało sens.
Obecnie zostawiłem na samym `userId` i dopóki nie używam hibernate'a, to powinno być ok. W przypadku gdy zacznę tutaj używać hiberante'a to po dodaniu adnotacji `@Id` hash code dla każdego obiektu będzie musiał być taki sam [url](https://vladmihalcea.com/how-to-implement-equals-and-hashcode-using-the-jpa-entity-identifier/)

tomaszc commented on src/main/java/com/trzewik/jdbc/db/AccountDao.java:
```diff
@@ -0,0 +1,69 @@
+
+    @Override
+    public void save(@NonNull Account account) throws SQLException {
+        String insert = String.format("INSERT INTO %s.account (username, email) VALUES ('%s', '%s')",
```
zastanów się jaki typ ataku mogę w ten sposób przeprowadzić i jak go ominąć

Czeffik replied:
SQL injection :C fixed

tomaszc commented on src/main/java/com/trzewik/jdbc/db/AccountDao.java:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    void saveMany(List<Account> accounts) throws SQLException {
+        db.startTransaction();
```
czy jest sens startować transakcję jeżeli accounts jest puste? transakcja to bardzo cenny zasób

tomaszc commented on src/main/java/com/trzewik/jdbc/db/AccountDao.java:
```diff
@@ -0,0 +1,69 @@
+        db.executeUpdate(delete);
+    }
+
+    void saveMany(List<Account> accounts) throws SQLException {
```
wszędzie @NonNull a tu nie?:>

tomaszc commented on src/main/java/com/trzewik/jdbc/db/AccountDao.java:
```diff
@@ -0,0 +1,69 @@
+    void saveMany(List<Account> accounts) throws SQLException {
+        db.startTransaction();
+        try {
+            if (accounts != null && accounts.size() > 0) {
```
porównywanie cyferek w `accounts.isEmpty()`

tomaszc commented on src/main/java/com/trzewik/jdbc/db/Db.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    void startTransaction() throws SQLException {
+        connection.setAutoCommit(false);
```
to jest start transakcji? hmmm

Czeffik replied:
tak jakby? :D 
https://docs.oracle.com/javase/tutorial/jdbc/basics/transactions.html

tomaszc commented on src/test/commons/com/trzewik/jdbc/db/AccountCreator.groovy:
```diff
@@ -0,0 +1,26 @@
+        )
+    }
+
+    static class AccountBuilder{
```
jaki ten obiekt ma związek z design-patternem o nazwie `Builder` który jest w suffixie jego nazwy?

Czeffik replied:
zmieniam nazwę na `AccountCreator`. I tak, można to tę klasę zastąpić mapą przekazywaną do metody `createAccount(Map account)`, ale:
- nie będę mógł zmienić tylko jedne defaultowej wartości - albo biorę wszystko z defaultu, albo nic
- intelliJ nie będzie mi podpowiadał co tam mogę wstawić
- zmiana nazwy pola jest mniej bolesna w przypadku klasy niż mapy (łatwiej będzie to refaktorować)
- można mieć kilka konstruktorów, które będą tworzyć dany obiekt w różny sposób :D i będziemy mieli to wszystko w jednym miejscu

tomaszc commented on src/test/integration/com/trzewik/jdbc/db/AccountDaoIT.groovy:
```diff
@@ -0,0 +1,180 @@
+
+    def 'should get all accounts from db'() {
+        expect:
+        dao.getAll().size() == 0
```
what do yo uthink of:
```
given:
    assert dao.getAll().isEmpty() 
when:
   ...
```
?

tomaszc commented on src/test/integration/com/trzewik/jdbc/db/AccountDaoIT.groovy:
```diff
@@ -0,0 +1,180 @@
+    def 'should get account by user_id'() {
+        given:
+        def account = createAccount()
+        def userId = dbHelper.insert(account).first().first()
```
oh my, this is complicated. WTF is happening here? make it readable

tomaszc commented on src/test/integration/com/trzewik/jdbc/db/AccountDaoIT.groovy:
```diff
@@ -0,0 +1,180 @@
+        account = createAccount(userId, account)
+
+        expect:
+        dao.get(account.userId).get() == account
```
read about hibernate and difference between their `get` and `find` methods,

tomaszc commented on src/test/integration/com/trzewik/jdbc/db/AccountDaoIT.groovy:
```diff
@@ -0,0 +1,180 @@
+
+    def 'should return null when can not find account with user_id'() {
+        given:
+        dbHelper.deleteAccountsByIds([4])
```
is it safe to mess around with test data like that? other tests can rely on it?

tomaszc commented on src/test/integration/com/trzewik/jdbc/db/AccountDaoIT.groovy:
```diff
@@ -0,0 +1,180 @@
+        def userId = dbHelper.insert(account).first().first()
+
+        and:
+        def updatedAccount = createAccount(new AccountBuilder(
```
bleeee, is it readable? please provide some `.clone()` method

tomaszc commented on src/test/integration/com/trzewik/jdbc/db/AccountDaoIT.groovy:
```diff
@@ -0,0 +1,180 @@
+        def second = createAccount()
+
+        when:
+        (dao as AccountDao).saveMany([first, second])
```
why does it fail? describe somehow how it makes it fail because I have WTF why would it do so

tomaszc commented on src/test/integration/com/trzewik/jdbc/db/AccountDaoIT.groovy:
```diff
@@ -0,0 +1,180 @@
+        (dao as AccountDao).saveMany(ACCOUNTS)
+
+        then:
+        dbHelper.allAccounts.size() == 0
```
isEmpty()

tomaszc commented on src/test/integration/com/trzewik/jdbc/db/AccountDaoIT.groovy:
```diff
@@ -0,0 +1,180 @@
+    }
+
+    @Unroll
+    def 'should do nothing when accounts are: #ACCOUNTS'() {
```
`when no accounts are provided: #DESCRIPTION` ?

tomaszc reviewed (commented):
Good job! Tomorrow I will take a look at the other PullRequest, leave it there :D

Czeffik committed:
After code review changes

EqualsAndHashCode in Account now is created only from userId.
AccountDao using binding and PreparedStatement for preventing
sql injection by using String.format(). Method get() renamed
to find() going with standards - find should return Optional,
get() should return object and when not found should return null.
Renamed AccountBuilder to AccountCreator for prevent misleading
user with understanding what is it this class. Refactored tests
for simple checks in database using methods which return all
records from table, delete all records or insert record into db.

Czeffik committed:
Add AccountService which give user possibility to create/get/delete/update Account

Czeffik committed:
Add ui for interacting with database

Added possibility to interact with application using console.
User have possibility to create/get/delete/update accounts in
database using commands (AccountController.Action).

Czeffik committed:
Add in memory dao implementation

Czeffik committed:
Add possibility to create accounts using csv file

Created CsvParser class which is responsible for read csv file
and transform it to list with lines and arguments List<List<String>>.
Added FileReader interface and AccountCsvReader implementation of it
which is used for reading accounts from csv file.
Added possibility in AccountController to create Accounts typing
absolute path to csv file with accounts.

Czeffik committed:
Refactor parseLine method in CsvParser

Czeffik committed:
Add domain package with AccountService and AccountRepository

This refactoring gave possibility to remove all throws SQLException from
AccountService methods.
Added domain package with: Account, AccountService and AccountRepository.
Now AccountService using AccountRepository to interact with database
instead Dao<AccountEntity>. Introduced Account which is domain representation
of AccountEntity.
In db package removed InMemoryAccountDao and AccountFactory classes.
Added implementations of AccountRepository - normal (connect with real db)
and in memory (HashMap as db).
In all other packages start using Account instead of AccountEntity.

Czeffik committed:
Move CsvParser to reader package

Czeffik committed:
Rename AccountController to AccountConsoleUI

AccountController in java world ia associated with REST controller.
Switch tricky class name to more meaningful.

Czeffik committed:
Rename FileReader interface to AccountProvider and move to domain
